sdk/cosmos/azure_data_cosmos/src/retry_policies/mod.rs

```diff
@@ -0,0 +1,319 @@
+            back_off_time: Duration::ZERO,
+        }
+    }
+
+    /// Creates a result indicating the operation should be retried after a delay
+    ///
+    /// # Arguments
+    /// * `backoff` - The duration to wait before retrying
+    ///
+    /// # Example
+    /// ```
+    /// use azure_data_cosmos::retry_policies::ShouldRetryResult;
+    /// use std::time::Duration;
+    ///
+    /// let result = ShouldRetryResult::retry_after(Duration::from_secs(5));
+    /// assert!(result.should_retry);
+    /// assert_eq!(result.back_off_time, Duration::from_secs(5));
+    /// ```
+    pub fn retry_after(backoff: Duration) -> Self {
+        Self {
+            should_retry: true,
+            back_off_time: backoff,
+        }
+    }
+}
+
+/// Trait defining the retry policy interface for Cosmos DB operations
+///
+/// This trait provides a contract for implementing retry logic for transient failures
+/// in Azure Cosmos DB operations. Implementers can define custom retry behavior for
+/// both exceptions (errors) and HTTP responses based on their specific requirements.
+#[async_trait]
+pub trait RetryPolicy: Send + Sync {
+    /// Called before sending a request to allow policy-specific modifications
+    ///
+    /// This method is invoked immediately before each request is sent (including retries).
+    /// # Arguments
+    /// * `request` - Mutable reference to the HTTP request being sent
+    fn on_before_send_request(&self, request: &mut Request);
+
+    /// Determines whether to retry an operation that resulted in an error
+    ///
+    /// This method is called when an operation fails with an exception (network error,
+    /// timeout, or HTTP error). The implementation should examine the error and decide
+    /// whether the operation should be retried and how long to wait before retrying.
+    ///
+    /// # Arguments
+    /// * `err` - The error that occurred during the operation
+    ///
+    /// # Returns
+    /// `ShouldRetryResult` indicating:
+    /// - `should_retry`: Whether the operation should be retried
+    /// - `backoff_time`: How long to wait before retrying (if `should_retry` is true)
+    async fn should_retry_exception(&self, err: &azure_core::Error) -> ShouldRetryResult;
+
+    /// Determines whether to retry an operation based on the HTTP response
+    ///
+    /// This method is called when an operation completes with an HTTP response that
+    /// might indicate a transient failure (e.g., 429, 503, 410). The implementation
+    /// should examine the response status code and headers to decide whether to retry.
+    ///
+    /// # Arguments
+    /// * `response` - The HTTP response received from the server
+    ///
+    /// # Returns
+    /// `ShouldRetryResult` indicating:
+    /// - `should_retry`: Whether the operation should be retried
+    /// - `backoff_time`: How long to wait before retrying (if `should_retry` is true)
+    async fn should_retry_response(&self, response: &RawResponse) -> ShouldRetryResult;
+}
```

Comment:
I think this structure of a before send/should retry method is likely to get us in to state-management problems.  The `Policy`s used in `azure_core` for the standard HTTP pipeline use a fairly simple middleware pattern that I think would be more appropriate for us to use here. In those, each policy is given the request and expected to return the response. But in order to do that, they are also given a list of all the remaining policies in the pipeline so they can "pass along" the request to the next policy.

This does basically mean embedding the `RetryHandler` logic to handle sleeping during backoff in each policy, but we can solve that duplication by writing helper functions that take care of the common logic.